### ISSUE TITLE
Bad macro usage error message does not include correct error location

### ISSUE THREAD

**Issue by SimonSapin** [rust-lang/rust#31022]
Consider this reduced test case:

```rust
macro_rules! match_ignore_ascii_case {
    (@inner $value:expr) => { () };
    ( $($rest:tt)* ) => { match_ignore_ascii_case!(@inner $($rest)*) };
}

fn main() {
    // This is fine
    match_ignore_ascii_case!(1);

    // This causes an error as it doesn’t match the expected syntax
    // but the error message does not the location of the actual error.
    match_ignore_ascii_case!(2 => 3);
}
```

The `@inner` indirection exists because the non-reduced macro is recursive:
* https://play.rust-lang.org/?gist=f8b1652f43cc720f89a3&version=nightly
* https://users.rust-lang.org/t/writing-a-macro-rules-macro-used-like-a-match-expression/4328

This fails to compile (as it should) but the error message does not include the real location of the error, which is line 12. It can be hard to track down in a large crate with many users of the macro.

```rust
a.rs:3:52: 3:53 error: unexpected token: `@`
a.rs:3     ( $($rest:tt)* ) => { match_ignore_ascii_case!(@inner $($rest)*) };
```

The error message looks even worse when the macro is used (with incorrect syntax) from another crate

```
<cssparser macros>:12:1: 12:2 error: unexpected token: `@`
<cssparser macros>:12 @ inner $ value , ( $ ( $ rest ) * ) -> (
```

**Comment by durka** [comment]
When the expanded code contains a non-macro-related error (say, a type mismatch), it at least shows a macro expansion backtrace. It's very hard to read, but gives spans at least. I guess that would be a start.

**Comment by durka** [comment]
The trouble is you get the _same_ error if you screw up writing the macro, e.g. misspell `@inner` or order something backwards in one of the arms. It isn't the caller's fault for sure.

**Comment by SimonSapin** [comment]
Yes, I don’t have an example at hand to copy/past but I see the kind of error message you mean. It *may* be the caller’s fault even if it isn’t necessarily. A macro expansion backtrace would be useful here.

**Comment by estebank** [comment]
Current output:

```
error: recursion limit reached while expanding the macro `match_ignore_ascii_case`
  --> src/main.rs:3:27
   |
3  |     ( $($rest:tt)* ) => { match_ignore_ascii_case!(@inner $($rest)*) };
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
12 |     match_ignore_ascii_case!(2 => 3);
   |     --------------------------------- in this macro invocation
   |
   = help: consider adding a `#![recursion_limit="128"]` attribute to your crate
```

**Comment by estebank** [comment]
Triage: no change.

**Comment by estebank** [comment]
Triage: no change.

**Comment by estebank** [comment]
Triage: [no change](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d2a14257ba4d13824bdbc6b73fb79d8c).

**Comment by estebank** [comment]
@SimonSapin do you have any recollection of this ticket? I am tempted to close (we now do point at line 12 after all) and I'm not sure how we can improve it. Do you feel we should keep this ticket around?

**Comment by SimonSapin** [comment]
I don’t remember this but based given that I wrote:

> the error message does not include the real location of the error, which is line 12

… and that the new message does point to line 12, I think we can close this as fixed. Looks like we could even have closed in 2017, sorry I didn’t realize!
